tests/geom/which_node_am_i_test.C

```diff
@@ -0,0 +1,155 @@
+// warnings.  These can be ignored in GCC as long as we wrap the
+// offending code in appropriate pragmas.  We can't get away with a
+// single ignore_warnings.h inclusion at the beginning of this file,
+// since the libmesh headers pull in a restore_warnings.h at some
+// point.  We also don't bother restoring warnings at the end of this
+// file since it's not a header.
+#include <libmesh/ignore_warnings.h>
+
+using namespace libMesh;
+
+class WhichNodeAmITest : public CppUnit::TestCase
+{
+
+public:
+  CPPUNIT_TEST_SUITE( WhichNodeAmITest );
+  CPPUNIT_TEST( testPyramids );
+  CPPUNIT_TEST( testPrisms );
+  CPPUNIT_TEST( testTets );
+  CPPUNIT_TEST( testHexes );
+  CPPUNIT_TEST_SUITE_END();
+
+public:
+
+  /**
+   * Note: the sections below marked
+   *
+   * #if !defined(NDEBUG) && defined(LIBMESH_ENABLE_EXCEPTIONS)
+   *
+   * are currently commented out because (even though they work) they
+   * print a bunch of stuff to the screen (error message, stack trace,
+   * etc.) that makes it look like they failed. If we could silently
+   * avoid seeing all that in the output stream, it might be worth
+   * including this test. The Google Test suite may provide a better
+   * mechanism for doing this.
+   */
+
+  void testPyramids()
+  {
+    // The last node on the right side (1) should be node 4 (apex node).
+    const Elem & pyr5 = ReferenceElem::get(PYRAMID5);
+    CPPUNIT_ASSERT_EQUAL(static_cast<unsigned int>(4), pyr5.which_node_am_i(/*side=*/1, /*node=*/2));
+
+// #if !defined(NDEBUG) && defined(LIBMESH_ENABLE_EXCEPTIONS)
+//     try
+//       {
+//         // Asking for the 4th node on a triangular face should throw.
+//         unsigned int n = pyr5.which_node_am_i(1, 3);
+//
+//         // We shouldn't get here if the line above throws. If we do
+//         // get here, there's no way this assert will pass.
+//         CPPUNIT_ASSERT_EQUAL(static_cast<unsigned int>(-1), n);
+//       }
+//     catch (...) {}
+// #endif
+
+#ifdef NDEBUG
+    // In optimized mode, we expect to get the "dummy" value 99.
+    unsigned int n = pyr5.which_node_am_i(1, 3);
+    CPPUNIT_ASSERT_EQUAL(static_cast<unsigned int>(99), n);
+#endif
+
+    // The last node on the right side (1) should be node 10.
+    const Elem & pyr13 = ReferenceElem::get(PYRAMID13);
+    CPPUNIT_ASSERT_EQUAL(static_cast<unsigned int>(10), pyr13.which_node_am_i(/*side=*/1, /*node=*/5));
+
+    // The central node of the base should be node 13
+    const Elem & pyr14 = ReferenceElem::get(PYRAMID14);
+    CPPUNIT_ASSERT_EQUAL(static_cast<unsigned int>(13), pyr14.which_node_am_i(/*side=*/4, /*node=*/8));
+  }
+
+
+
+  void testPrisms()
+  {
+    // A PRISM6 has four nodes on some sides and three nodes on others
+    const Elem & prism6 = ReferenceElem::get(PRISM6);
+    CPPUNIT_ASSERT_EQUAL(static_cast<unsigned int>(4), prism6.which_node_am_i(/*side=*/4, /*node=*/1));
+
+// #if !defined(NDEBUG) && defined(LIBMESH_ENABLE_EXCEPTIONS)
+//     try
+//       {
```

Comment:
> Also, I need to figure out how to debug broken unit tests properly. For applications I usually just set a breakpoint at MPI_Abort, but that doesn't work with unit tests that catch errors and continue. For our unit tests I used to "catch throw", but ever since we added the convergence_failure catch with pyramids I have to comment those FE tests out to avoid false positives, which is obviously a hack that doesn't scale.

I know there was some discussion of Google Test the other day, but I came across [Catch](https://github.com/philsquared/Catch) the other day (I think I saw @dmcdougall starred it, but not sure if he's actually tried it), but one of the neat looking things about it is that it looks like you can defines "modules" within the test suite and then only run those modules. I could see that being helpful on this point, e.g. only running `DistributedMesh` tests, or some such. But maybe Google Test has a similar feature?